// datakund/GetData.js
var ip = require("ip");
var http = require("http");
var execFile = require('child_process').execFile, child;
const request = require('request');
const findChromeVersion = require("find-chrome-version")
function changebotname(bot,domain){
var prev=''
var newname=''
for (chrs in bot) {
  var chr=bot[chrs]
  if(prev=='_'){
	newname=newname+chr.charAt(0).toUpperCase();
}else{
	newname=newname+chr
}
prev=chr
}
newname=newname[0].toUpperCase() + newname.slice(1);
if(domain==""){
}
else{
	newname=domain+"_"+newname
}
newname=newname[0].toUpperCase() + newname.slice(1);
return newname
}

function gettheuserid(){
	userid=ip.address()
	userid = userid.split('.').join('-')
	return userid
}
function getversion(calback){ 
  const ipaddress = async () => {
  const chromeVersion = await findChromeVersion()
  var version=chromeVersion
  version=version.split(".")
  version=version[0]
  calback(version)
}
ipaddress()
}
function handle_exc(error,stdout,stderr){
            if (error) {
				execFile('./node_modules/datakund/DataKund.exe', function(error,stdout,stderr){
					if (error) {
					}
					else{
						console.log("Started Application!!!!")
					}
				})
              }
			else{
				console.log("Started Application!!!!")
			}
}
function start_exe(){
	child = execFile('DataKund.exe', handle_exc); 
}
function call_request(options,cbk2){
	request(options, (err, res, body) => {
	  if (err) { 
		  setTimeout(function(){call_request(options,cbk2)}, 1000)
		  }
	  else{
		  cbk2()
		  done=true
	  }
    });
}
function while_loop(options,cbk2){
	var i=0
	var done=false
    setTimeout(function(){call_request(options,cbk2)}, 1000)
	//cbk2()
}
function makerequest(post_data,cbk1,cbk2){
	var jsonDataObj =post_data
	jsonDataObj["tech_type"]="NPMJS"
    var options = {"method": "POST","url": "http://127.0.0.1:5350/install","headers": {"content-type": "application/json",},"body": jsonDataObj,"json": true};
	var done=false
   request(options, (err, res, body) => {
  if (err) {
      start_exe()
      }
  else{
      done=true}
	  while_loop(options,cbk2)
      });
	
}
function checkifinstall(calback){
	userid=gettheuserid()
	getversion(function(version){
		makerequest({"userid":userid,"majorVersion":version,"browser":"chrome"},function(){},calback)
	})
}

module.exports = { changebotname ,gettheuserid,getversion,checkifinstall}
